File: cleannb/clean_nb.py
from json import dump, load
from shutil import copyfile

from .clean_jl import clean_jl
from .clean_py import clean_py
from .has_julia_and_juliaformatter import has_julia_and_juliaformatter


def clean_nb(pa, ne):

    if ne:

        pa = copyfile(pa, pa.replace(".ipynb", ".clean.ipynb"))

    with open(pa) as io:

        nb = load(io)

    if "language_info" in nb["metadata"]:

        la = nb["metadata"]["language_info"]["name"]

        if la == "ipython" or "colab" in nb["metadata"]:

            la = "python"

    else:

        la = ""

    if la == "julia" and has_julia_and_juliaformatter():

        fu = clean_jl

    elif la == "python":

        fu = clean_py

    else:

        fu = None

    ce_ = []

    for ce in nb["cells"]:

        if "execution_count" in ce:

            ce["execution_count"] = None

        if "outputs" in ce:

            ce["outputs"] = []

        if "jupyter" in ce["metadata"] and "source_hidden" in ce["metadata"]["jupyter"]:

            ce["metadata"]["jupyter"].pop("source_hidden")

        if ce["cell_type"] == "code" and fu is not None:

            co = "".join(ce["source"])

            if co.strip() == "":

                continue

            ce["source"] = fu(co).splitlines(True)

        ce_.append(ce)

    nb["cells"] = ce_

    with open(pa, "w") as io:

        dump(nb, io, indent=1)

        io.write("\n")

File: src/clean_nb.jl
function clean_nb(pa; ke_ar...)

    println("Formatting ", pa)

    nb = OnePiece.dict.read(pa)

    me = nb["metadata"]

    ju = me["language_info"]["name"] == "julia"

    if ju

        me["language_info"]["version"] = string(VERSION)

        me["kernelspec"]["name"] = "julia-$(VERSION.major).$(VERSION.minor)"

    else

        println("Language is not julia.")

    end

    ce_ = []

    for ce in nb["cells"]

        if ce["cell_type"] == "code"

            co = join(ce["source"])

            if isempty(strip(co))

                continue

            end

            ce["execution_count"] = nothing

            ce["outputs"] = []

            if ju

                li_ = split(JuliaFormatter.format_text(co; ke_ar...), "\n")

                ce["source"] = [string.(li_[1:(end - 1)], "\n"); li_[end]]

            end

            ce["metadata"] = Dict()

        end

        push!(ce_, ce)

    end

    nb["cells"] = ce_

    nb = OnePiece.dict.sort_recursively!(nb)

    OnePiece.dict.write(pa, nb, id = 1)

end
